crypto/src/merkle.rs

```diff
@@ -189,173 +189,142 @@ impl<T> Node<T> {
     }
 
     #[cfg(feature = "std")]
-    fn nodes_pair_hash(left: &Self, right: &Self) -> HashOf<Self> {
-        let left_hash = left.hash();
-        let right_hash = right.hash();
+    fn nodes_pair_hash(left: &Self, right: &Self) -> Option<HashOf<Self>> {
+        let (left_hash, right_hash) = match (left.hash(), right.hash()) {
+            (None, None) => return None,
+            (None, Some(r_hash)) => return Some(r_hash),
+            (Some(l_hash), None) => return Some(l_hash),
+            (Some(l_hash), Some(r_hash)) => (l_hash, r_hash),
+        };
         let sum: Vec<_> = left_hash
             .as_ref()
             .iter()
             .zip(right_hash.as_ref().iter())
             .map(|(l, r)| l.saturating_add(*r))
             .collect();
-        crate::Hash::new(sum).typed()
+        Some(crate::Hash::new(sum).typed())
+    }
+
+    fn children(&self) -> Option<[&Self; 2]> {
+        if let Node::Subtree(subtree) = self {
+            return Some([&*subtree.left, &*subtree.right]);
+        }
+        None
     }
 }
 
-impl<'itm, T> BreadthFirstIter<'itm, T> {
-    fn new(root_node: &'itm Node<T>) -> Self {
-        BreadthFirstIter {
-            queue: vec![root_node],
+impl<'itm, T> BreadthFirstNodeIterator<'itm, T> {
+    #[inline]
+    fn new(root: &'itm Node<T>) -> Self {
+        Self {
+            queue: VecDeque::from(vec![root]),
```

Comment:
another nitpick, you could determine the max size of the queue and use `VecDeque::with_capacity`

Reply:
Addressed in 6c93c3c6992dac9e4c880f2a87d83c9b1e089f6d

Reply:
Do you think we should use DFS to iterate over nodes/leaves in the first place?
In that case, we would use `log(n_leaves)` capacity of `Vec` instead of `n_leaves` capacity of `VecDeque`.
However, this does not affect the tree structure and could be commented out and left to another optimizing PR